modules/gapi/src/streaming/onevpl/demux/async_mfp_demux_data_provider.cpp

```diff
@@ -0,0 +1,783 @@
+                               worker_key_to_buffer_mapping_storage.size());
+        for (auto& buffer : worker_key_to_buffer_mapping_storage) {
+            if (buffer.second) {
+                buffer.second->Unlock();
+            }
+        }
+        worker_key_to_buffer_mapping_storage.clear();
+    }
+
+    GAPI_LOG_INFO(nullptr, "Clean didn't used up storage, count: " <<
+                           processing_key_to_buffer_mapping_storage.size());
+    for (auto& buffer : processing_key_to_buffer_mapping_storage) {
+        if (buffer.second) {
+            buffer.second->Unlock();
+        }
+    }
+    processing_key_to_buffer_mapping_storage.clear();
+
+    // release COM object before overall MFP shutdown
+    source_reader.reset();
+    source.reset();
+
+    GAPI_DbgAssert(com_interface_reference_count.load() > 0 &&
+                   "Incorrect reference counting for MFPAsyncDemuxDataProvider");
+    MFShutdown();
+    GAPI_LOG_INFO(nullptr, "[" << this << "] " <<
+                           "deinitialized");
+}
+
+
+ULONG MFPAsyncDemuxDataProvider::AddRef() {
+    // align behavior with InterlockedIncrement
+    return com_interface_reference_count.fetch_add(1) + 1;
+}
+
+ULONG MFPAsyncDemuxDataProvider::Release() {
+    auto count = com_interface_reference_count.fetch_sub(1);
+    GAPI_Assert(count != 0 && "Incorrect reference counting for MFPAsyncDemuxDataProvider");
+    count -= 1; // align behavior with InterlockedDecrement
+    return count;
+}
+
+HRESULT MFPAsyncDemuxDataProvider::QueryInterface(REFIID riid, void** ppv)
+{
+    static const QITAB qit[] =
+    {
+        QITABENT(MFPAsyncDemuxDataProvider, IMFSourceReaderCallback),
+        { 0 },
+    };
+    return QISearch(this, qit, riid, ppv);
+}
+
+
+STDMETHODIMP
+MFPAsyncDemuxDataProvider::OnReadSample(HRESULT status, DWORD stream_index,
+                                        DWORD stream_flag, LONGLONG timestamp,
+                                        IMFSample *sample_ptr) {
+    return on_read_sample_impl(status, stream_index, stream_flag,
+                               timestamp, sample_ptr);
+}
+
+HRESULT MFPAsyncDemuxDataProvider::on_read_sample_impl(HRESULT status, DWORD,
+                                                       DWORD stream_flag, LONGLONG,
+                                                       IMFSample *sample_ptr) {
+    GAPI_LOG_DEBUG(nullptr, "[" << this << "] status: " << std::to_string(HRESULT_CODE(status)) <<
+                            ", stream flags: " << stream_flag <<
+                            ", sample: " << sample_ptr);
+    HRESULT hr = S_OK;
+    if (FAILED(status)) {
+        hr = status;
+    }
+
+    // check EOF
+    if (stream_flag & MF_SOURCE_READERF_ENDOFSTREAM) {
+        GAPI_LOG_DEBUG(nullptr, "[" << this << "] EOF");
+
+        // close reader
+        provider_state.store(State::Exhausted);
+        buffer_storage_non_empty_cond.notify_all();
```

Comment:
Will `notify_one` work here?

Reply:
Here and below

Reply:
I don't think more than one thread should wake in your case, since others will be still blocked if I get it correctly. There is no major difference though. Feel free to resolve this comment